chain/chain/src/store.rs

```diff
@@ -1734,14 +1735,47 @@ impl<'a> ChainStoreUpdate<'a> {
             .insert((height, shard_id), chunk_hash);
     }
 
-    pub fn clear_old_data_on_height(&mut self, height: BlockHeight) -> Result<(), Error> {
+    pub fn clear_old_data_on_height(
+        &mut self,
+        trie: Arc<Trie>,
+        height: BlockHeight,
+    ) -> Result<(), Error> {
         let mut store_update = self.store().store_update();
         let blocks_current_height = match self.get_all_block_hashes_by_height(height) {
             Ok(blocks_current_height) => {
                 blocks_current_height.values().flatten().cloned().collect()
             }
             _ => vec![],
         };
+
+        // 1. Trie cleaning
+        if let Ok(canonical_hash) = self.get_block_hash_by_height(height) {
+            for block_hash in blocks_current_height.iter() {
+                if *block_hash == canonical_hash {
+                    // Block on canonical chain must be processed after all
+                    continue;
+                }
+                // 1a. Apply revert insertions from ColTrieChanges, cleaning forks
+                self.store()
+                    .get_ser(ColTrieChanges, block_hash.as_ref())?
+                    .map(|trie_changes: TrieChanges| {
+                        trie_changes
+                            .revert_insertions_into(trie.clone(), &mut store_update)
+                            .map_err(|err| ErrorKind::Other(err.to_string()))
+                    })
+                    .unwrap_or_else(|| Ok(()))?;
+            }
+            // 1b. Apply deletions from ColTrieChanges, cleaning main chain
```

Comment:
this should be in an else, applying deletions from forks would definitely trigger some asserts if we had tests